In [1]:
import pandas as pd
import boto3
import os 
import datetime as dt
import time
import gzip
import json
import io

In [2]:
# Manual for now
KEY="AKIA3E4BU4DC3XNHFAUH"
SECRET='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'
S3_Bucket='earnest-spend-279-us-east-1'
S3_Prefix='query/v1.0/'
S3_Path = 's3://earnest-spend-279-us-east-1/query/v1.0/YYYY/MM/DD/'

In [3]:
test=pd.DataFrame({"Param":
                  ["KEY", "SECRET", "S3_Bucket", "S3_Prefix", "S3_Path"],
              "Value":
                  [KEY, SECRET, S3_Bucket, S3_Prefix, S3_Path]
             })

s3 = boto3.resource('s3',
                   region_name = 'us-east-1',
                   aws_access_key_id=KEY,
                    aws_secret_access_key=SECRET)


s3_client = boto3.client('s3',
                         aws_access_key_id=KEY,
                         aws_secret_access_key=SECRET)

In [57]:
# Most recent business day and Thursday
last_bus='2021/06/01'
last_thurs = '2021/05/27'

In [59]:
bucket=s3.Bucket(S3_Bucket) #S3_Bucket  'earnest-spend-279-us-east-1'
data = [filename.key for filename in bucket.objects.filter(Prefix=S3_Prefix)]
print("There are a total of","{:,}".format(len(data)),"files.")
data_thurs = [filename.key for filename in bucket.objects.filter(Prefix="query/v1.0/"+last_thurs)]  # update to latest thursday
data_mon = [filename.key for filename in bucket.objects.filter(Prefix="query/v1.0/"+last_bus)]    # update to latest monday
print("On Thursdays, there are ", "{:,}".format(len(data_thurs)),"files, and on other days there are","{:,}".format(len(data_mon)),"files.")

There are a total of 16,851 files.
On Thursdays, there are  3,222 files, and on other days there are 1,301 files.


##### From: Matthew Crossley Colbeck <mcolbeck@earnestresearch.com>

To start I recommend focusing on the <b> KPISpend file - machine_data_quarter </b> here you'll find <200 core covered merchants with backtests vs. reported revenues.

(Note: he is referring to <b> machine_data_quarter</b>  and <b>machine_data_daily </b> for KPI Spend

For broader coverage of ~3,000 public/private merchants use <b> QuerySpend </b> which has a host of tables to address a number of uses. Here I suggest starting with <b> machine_data_quarter</b>  which as with the KPI file uses the correct fiscal company calendar (albeit w/o a company reported metric).

(Note: for Query Spend there are a bunch of tables (see the excel file he sent)

##### Tables:

* change_in_spend
* change_in_spend_channel
* cohort_retention
* consensus
* cross_shop
* cross_shop_channel
* cs_panel_spend
* customer_metrics
* machine_data_daily
* machine_data_daily_enhanced
* machine_data_quarter
* quintile_cohort
* subscription_cohort
* subscription_machine_data_daily_enhanced
* ticker_to_merchant_mapping
* ticket_frequency_daily
* earnest_opinions_YYYYMMDD

For <b> custom date aggregations </b> and to <b>compare companies on different fiscal calendars in a heatmap</b> or similar we suggest using the <b>machine_data_daily</b> or possibly __machine_data_daily_enhanced__ if you wanted to include the channel attribute (in-store/online/store-card) although you may want to sum up cities for a national read.

#### Summary from AWS Queries:
~1301 files total for latest day...
~3223 files for each thursday, captures the weekly only files

* machine_data_quarter - 1 file
* machine_data_daily - 1 file
* machine_data_daily_enhanced -1280 files
* change_in_spend - 192 
* change_in_spend_channel - 192
* consensus 1
* cross shop -192
* cross shop channel -192
* cohort 192
* cs_panel 1
* customer 192
* quintile 192
* subscription_cohort 192
* subscription_machine daily enhanced 192
* ticker_to_merchant_mapping 1
* ticket_frequency_daily 192
* earnest_opinions_YYYYMMDD ??

In [60]:
# Daily Consensus File
consensus = [filename.key for filename in bucket.objects.filter(Prefix="query/v1.0/" + last_bus+"/consensus")] 
print("There is",len(consensus),"consensus file - name is...",consensus[0])

There is 1 consensus file - name is... query/v1.0/2021/06/01/consensus/consensus.gz


In [65]:
# Generic File...with only one item
query_root = "consensus"
gz_file = [filename.key for filename in bucket.objects.filter(Prefix="query/v1.0/"+last_bus +"/"+query_root)][0]

In [66]:
response = s3_client.get_object(Bucket=S3_Bucket, Key=gz_file)
status=response.get("ResponseMetadata",{}).get("HTTPStatusCode")

In [84]:
def download_pandas_gz(s3_client, bucket, key):
    ''' download gzipped json file from s3 and convert to dict '''
    
    response = s3_client.get_object(Bucket=bucket, Key=key)
    content = response['Body'].read()
    with gzip.GzipFile(fileobj=io.BytesIO(content), mode='rb') as fh:
        return pd.read_table(fh,"|")

In [85]:
df=download_pandas_gz(s3_client,bucket=S3_Bucket,key = gz_file)

In [69]:
df.columns

Index(['delivery_date', 'tab', 'label', 'reported_metric',
       'consensus_backtest', 'is_implied', 'implied_from_metric',
       'implied_from_consensus', 'spread'],
      dtype='object')

In [70]:
df.head()

delivery_date    tab       label                          reported_metric  \
0    2021-06-01    AAP        1Q21                                    Sales   
1    2021-06-01    ACI  Prelim1Q20              Net Sales and Other Revenue   
2    2021-06-01    ACI        1Q20              Net Sales and Other Revenue   
3    2021-06-01  ADRNY  Prelim2Q21                             US Net Sales   
4    2021-06-01    ADT  Prelim2Q21  Monitoring and Related Services Revenue   

   consensus_backtest  is_implied implied_from_metric  implied_from_consensus  \
0               0.224       False                 NaN                     NaN   
1                 NaN       False                 NaN                     NaN   
2                 NaN       False                 NaN                     NaN   
3                 NaN       False                 NaN                     NaN   
4              -0.013        True               Sales                  -0.033   

   spread  
0     NaN  
1     NaN  
2     NaN  
3     NaN  
4    0.02

In [77]:
ticker = "AAP"
ticker_data=df[df['tab']==ticker]
ticker_data

delivery_date  tab label reported_metric  consensus_backtest  is_implied  \
0    2021-06-01  AAP  1Q21           Sales               0.224       False   

  implied_from_metric  implied_from_consensus  spread  
0                 NaN                     NaN     NaN

In [89]:
df

delivery_date       tab       label  \
0      2021-06-01       AAP        1Q21   
1      2021-06-01       ACI  Prelim1Q20   
2      2021-06-01       ACI        1Q20   
3      2021-06-01     ADRNY  Prelim2Q21   
4      2021-06-01       ADT  Prelim2Q21   
..            ...       ...         ...   
236    2021-06-01  YUM-HABT  Prelim2Q21   
237    2021-06-01   YUM-KFC  Prelim2Q21   
238    2021-06-01    YUM-PH  Prelim2Q21   
239    2021-06-01    YUM-TB  Prelim2Q21   
240    2021-06-01      ZOES  Prelim2Q21   

                                   reported_metric  consensus_backtest  \
0                                            Sales               0.224   
1                      Net Sales and Other Revenue                 NaN   
2                      Net Sales and Other Revenue                 NaN   
3                                     US Net Sales                 NaN   
4          Monitoring and Related Services Revenue              -0.013   
..                                             ...                 ...   
236                Habit Division Systemwide Sales                 NaN   
237        KFC Division US Systemwide Sales Growth               0.318   
238  Pizza Hut Division US Systemwide Sales Growth               0.062   
239     Taco Bell Division Systemwide Sales Growth               0.187   
240                               Restaurant Sales                 NaN   

     is_implied         implied_from_metric  implied_from_consensus  spread  
0         False                         NaN                     NaN     NaN  
1         False                         NaN                     NaN     NaN  
2         False                         NaN                     NaN     NaN  
3         False                         NaN                     NaN     NaN  
4          True                       Sales                  -0.033    0.02  
..          ...                         ...                     ...     ...  
236       False                         NaN                     NaN     NaN  
237        True        KFC Same Store Sales                   0.278    0.04  
238        True  Pizza Hut Same Store Sales                   0.102   -0.04  
239        True  Taco Bell Same Store Sales                   0.147    0.04  
240       False                         NaN                     NaN     NaN  

[241 rows x 9 columns]

In [116]:
unique_metrics=list(set(list(df['reported_metric'].values)))
len(unique_metrics)

171

In [ ]:
# for files with multiple file parts per day

In [135]:
# Daily Cohort File - multiple files, each thursday
cohort = [filename.key for filename in bucket.objects.filter(Prefix="query/v1.0/" + last_thurs+"/cohort")] 
print("There are",len(cohort),"cohort files - names are...")
print("",cohort[0],"\n",cohort[1],"\n",cohort[2],"\n",cohort[3],"\n",cohort[4],"\n")
#consensus = [filename.key for filename in bucket.objects.filter(Prefix="query/v1.0/" + last_bus+"/consensus")] 


There are 192 cohort files - names are...
 query/v1.0/2021/05/27/cohort_retention/part-00000-3f4b4a1f-c551-4df8-b959-0f61280994ff-c000.csv.gz 
 query/v1.0/2021/05/27/cohort_retention/part-00001-3f4b4a1f-c551-4df8-b959-0f61280994ff-c000.csv.gz 
 query/v1.0/2021/05/27/cohort_retention/part-00002-3f4b4a1f-c551-4df8-b959-0f61280994ff-c000.csv.gz 
 query/v1.0/2021/05/27/cohort_retention/part-00003-3f4b4a1f-c551-4df8-b959-0f61280994ff-c000.csv.gz 
 query/v1.0/2021/05/27/cohort_retention/part-00004-3f4b4a1f-c551-4df8-b959-0f61280994ff-c000.csv.gz 



In [152]:
cohort0=download_pandas_gz(s3_client,bucket=S3_Bucket,key = cohort[0])
cohort1=download_pandas_gz(s3_client,bucket=S3_Bucket,key = cohort[191])

In [143]:
cohort0[cohort0['merchant']=='&pizza']

merchant      cohort cohort_type  periods_from_first_purchase  \
0      &pizza  2017-01-01       month                            0   
1      &pizza  2017-01-01       month                            1   
2      &pizza  2017-01-01       month                            2   
3      &pizza  2017-01-01       month                            3   
4      &pizza  2017-01-01       month                            4   
...       ...         ...         ...                          ...   
1491   &pizza  2021-04-01       month                            0   
1492   &pizza  2021-04-01       month                            1   
1493   &pizza  2021-04-01       month                            2   
1494   &pizza  2021-05-01       month                            0   
1495   &pizza  2021-05-01       month                            1   

      fully_aged_period    end_date  member_count  transaction_count  \
0                  True  2021-05-21           985             1004.0   
1                  True  2021-05-21           122              158.0   
2                  True  2021-05-21            97              128.0   
3                  True  2021-05-21            95              134.0   
4                  True  2021-05-21            87              112.0   
...                 ...         ...           ...                ...   
1491               True  2021-05-21           543              556.0   
1492              False  2021-05-21            79              100.0   
1493              False  2021-05-21            13               14.0   
1494              False  2021-05-21           351              359.0   
1495              False  2021-05-21            15               15.0   

      panel_sales  
0        16026.18  
1         2448.57  
2         2125.92  
3         2120.56  
4         1657.51  
...           ...  
1491     12743.89  
1492      1833.45  
1493       198.49  
1494      7556.35  
1495       200.62  

[1496 rows x 9 columns]

In [153]:
cohort1

merchant      cohort cohort_type  periods_from_first_purchase  \
0            Zazzle  2021-01-01     quarter                            2   
1            Zazzle  2021-04-01     quarter                            0   
2            Zazzle  2021-04-01     quarter                            1   
3              Zeel  2017-01-01     quarter                            0   
4              Zeel  2017-01-01     quarter                            1   
...             ...         ...         ...                          ...   
8845  zZounds Music  2021-01-01     quarter                            0   
8846  zZounds Music  2021-01-01     quarter                            1   
8847  zZounds Music  2021-01-01     quarter                            2   
8848  zZounds Music  2021-04-01     quarter                            0   
8849  zZounds Music  2021-04-01     quarter                            1   

      fully_aged_period    end_date  member_count  transaction_count  \
0                 False  2021-05-21            59               80.0   
1                 False  2021-05-21          3421             3467.0   
2                 False  2021-05-21           283              348.0   
3                  True  2021-05-21            69               79.0   
4                  True  2021-05-21            13               36.0   
...                 ...         ...           ...                ...   
8845               True  2021-05-21           298              306.0   
8846              False  2021-05-21           109              250.0   
8847              False  2021-05-21            36               59.0   
8848              False  2021-05-21            95               96.0   
8849              False  2021-05-21            21               31.0   

      panel_sales  
0         4910.89  
1       212056.63  
2        20020.33  
3         8163.47  
4         3867.90  
...           ...  
8845     64600.67  
8846     28706.39  
8847      5637.99  
8848     25766.19  
8849      4989.23  

[8850 rows x 9 columns]

In [156]:
list(set(list(cohort1['merchant'])))

['by CHLOE',
 'eMeals',
 'hhgregg',
 "Zoe's Kitchen (ZOES)",
 'iHeartMedia',
 'inkbox',
 'j2 Global (JCOM) - MyFax',
 'Zola',
 'doTERRA',
 'oVertone Haircare',
 'ezCater',
 'Zinus',
 'eSalon',
 'j2 Global (JCOM) - MetroFax',
 'j2 Global (JCOM)',
 'Zumper',
 'Zwift',
 'j2 Global (JCOM) - eFax',
 'minuteKEY',
 'honeygrow',
 'j2 Global (JCOM) - Fax.com',
 'Zesty Paws',
 'fuboTV (FUBO)',
 'sweetgreen',
 'Zeel',
 'Zenlen',
 'birddogs',
 'fresh &co',
 'eShakti',
 'wagamama',
 'Zenni Optical',
 'j2 Global (JCOM) - eVoice',
 'bebe',
 'goPuff',
 'Zoro Tools',
 'ePremium Insurance',
 'eBags.com',
 'eharmony',
 'eBay (EBAY)',
 'fromyouflowers.com',
 'tixr',
 'Zoom Video Communications (ZM)',
 'b8ta',
 'iRobot (IRBT)',
 'zZounds Music',
 'thredUP',
 'Zelle',
 'la Madeleine',
 'e.l.f. Cosmetics',
 'Zumiez (ZUMZ)',
 'Zillow Group (ZG)',
 'Zazzle',
 'Zynga (ZNGA)',
 'Ziply Fiber']